Add axis domain references to shapes, annotations, and layout images

nicholas-esterer opened the pull request:
Resolves #4958.

These elements can now be drawn relative to an axis's domain, so if this domain is updated, say, the element's position will update its position according to the new length and offset of the domain.
For example, to add a horizontal line that always spans the x axis and starts 25% of the way up the y axis, you would add such a shape to the layout:

``` javascript
{
          type: 'line',
          xref: 'x domain',
          yref: 'y domain',
          x0: 0,
          x1: 1,
          y0: 0.25,
          y1: 0.25,
}
```

You can see more examples of shapes by looking in `test/domain_ref_tests.html`. I've written a bunch of tests there that will eventually be made into jasmine tests (we can rebase out that file once that happens). I'll push some more tests (related to editing of shapes and annotations) as I make them.

nicholas-esterer commented:
<h1>Summary of this PR</h1>
<p>This PR makes it possible to specify:
- for shapes:
- the position in coordinates relative to the length or height of an
axis
- for annotations:
- the position in coordinates relative to the length or height of an
axis
- however the length of the arrow can only be specified in pixels or
absolute data coordinates
- for images:
- the position and size in coordinates relative to the length or height
of an axis</p>
<p>I wanted to document some of the shortcomings and discuss leaving them or their
eventual resoluion.</p>
<h2>Inconsistencies</h2>
<ul>
<li><code>src/components/shapes/defaults.js:handleShapeDefaults</code>
<ul>
<li>uses <code>var axRefInfo = Axes.extractInfoFromAxisRef(axRef)</code> then checks if
this equals 'domain'</li>
</ul>
</li>
<li><code>src/components/annotations/draw.js:drawRaw</code>
<ul>
<li>also does this</li>
</ul>
</li>
<li>but <code>src/components/images/draw.js:137</code> uses <code>d.xref.endsWith(' domain')</code></li>
<li><code>src/components/shapes/draw.js:setClipPath</code>
<ul>
<li>uses <code>/[xyz][1-9]* *domain/g</code></li>
</ul>
</li>
<li>but <code>src/lib/regex.js:27</code></li>
<li>suggests <code>/x'([2-9]|[1-9][0-9]+)?'/</code> which I changed (at (*)) in <code>src/components/annotations/attributes.js</code>:</li>
</ul>
<pre><code class="language-javascript">module.exports = templatedArray('annotation', {
    ...
    xref: values: (*)
    ...
    yref: values: (*)
    ...
}
</code></pre>
<p>to <code>cartesianConstants.AX_ID_EXTRAS_PATTERN.toString()</code> i.e.,
<code>/^[xyz]([2-9]|[1-9][0-9]+)?( domain)?$/</code></p>
<p>Some of these inconsistencies above are my own doing but I would like comments
on the best approach, e.g., is it better to check <code>.endsWith(' domain')</code> or use
a regular expression?</p>
<h2>repeated pixel calculations</h2>
<h3>In the following <code>files:functions</code></h3>
<pre><code>src/components/annotations/draw.js:drawRaw
src/components/images/draw.js:draw
src/components/shapes/draw.js:getPathString
src/components/shapes/helpers.js:exports.getDataToPixel
src/components/shapes/helpers.js:exports.getPixelToData
</code></pre>
<p>This was the state of the code when I started working on it, so I had to change
the code in these sections to be compatible with axis domain references. I am
presenting these parts here in case there is interest in obtaining some kind of
generalization, e.g., writing one function that will work for shapes, annotations,
and images.</p>
<p>The following might be familiar to you but is included here for reference. It is
a description of how the conversion to pixel coordinates happens for annotations
and compares it to that for shapes and images.</p>
<h2>How pixel conversion differs slightly:</h2>
<p><code>src/components/annotations/draw.js:drawRaw</code></p>
<ul>
<li>If ax specified, calculates pixel coordinate of arrowhed from data
coordinate but makes sure the arrowhead is onscreen.</li>
<li>If paper, calculates relative to <code>gd._fullLayout._size</code>.</li>
<li>If domain appended, calculates relative to axis size</li>
<li>for these last 2, if y axis, relative to top but complement of normalized
coordinate is taken and multiplied with height.
<code>p2rR2p</code>: this goes from pixel coordinates (coming from a <code>mousemove</code> event) to
the range relevant to the axis reference of that annotation (see
<code>helpers.js:exports.getPixelToData</code> for shapes)
After this, if <code>showarrow</code> is true, the position of the arrow is calculated
using <code>ax.r2p</code> if the <code>xref</code> is an axis only (e.g.,<code>'x'</code>)
Note however the use of <code>ax.r2p</code> here may be broken, we would
need to check by comparing an arrowlength specified in domain range
coordinates and one specified in pixels, that are trying to be equal (see
the comments on annotation <code>xref</code>, <code>yref</code> below to see why this doesn't
presently behave as (perhaps) expected).</li>
</ul>
<p><code>src/components/images/draw.js:draw</code>
For image width, some difference is used without offset to convert from
whatever coordinate specified into pixels, then for the actual image offset,
the offset is used to compute the absolute position. This is almost exactly
like for the arrowhead position of annotations. The width and height
calculations is analogous to the calculation of the arrow length in pixels
for annotations, which uses <code>ax.r2p</code>. However, the width and height
calculations for images to take into account the (WLOG) <code>xref='x'</code>, <code>xref='x domain'</code> and <code>xref='paper'</code> options (unlike the arrow length calculation).</p>
<p><code>src/components/shapes/draw.js:getPathString</code>
This is like the coordinate to pixel conversion done in
<code>src/components/annotations/draw.js:drawRaw</code> but here when converting to pixels by calling
<code>ax.r2p</code> this calls (if you dig down you'll see) <code>toLog</code> on the coordinate
(annotations don't do this but this is consistent with their positions
having to be specified as the log of the range (e.g., specifying 0.5 would
put you near the axis tick at ~3 because 10^0.5 ~ 3))</p>
<p><code>src/components/shapes/helpers.js:exports.getDataToPixel</code>
Basically the same as the primary pixel calculation for annotations (in
<code>src/components/annotations/draw.js:drawRaw</code>) except the one for annotations
checks if the annotation is offscreen and computes <code>autoAlignFraction</code> based
on the <code>showarrow</code> option.</p>
<p><code>src/components/shapes/helpers.js:exports.getPixelToData</code>
This is very similar to <code>p2rR2p</code> however, we can't just use one for both the
annotation and draw <code>moveFn</code>'s because the one for shapes calculates based on
the absolute pixel value, whereas the one for annotations uses the change in
pixel value. So generalizing this would require a bit of refactoring (one
would have to convert or not convert to absolute coordinates).</p>
<h2>annotation <code>axref</code>, <code>ayref</code></h2>
<p>WLOG here I discuss specifying the x position annotation text relative to
the x position of the annotation (the point of the arrow). It seems that
unless <code>axref === xref</code> then the position is calculated as an offset in pixels
from the annotation. Furthermore if you specify axref as anything other than
<code>/[xyz][2-9][0-9]*/</code> the position is still in pixels (I believe due to some
value coersion but I have not dug down to see if that is the case). So
<code>axref='x domain'</code> also specifying the position in pixels even if
<code>xref='x domain'</code> is consistent with
the old behaviour when <code>axref='paper'</code> and <code>xref='paper'</code>. So the question is:
for <code>'/[xyz][2-9][0-9]/ domain'</code>, do we keep the behaviour as for <code>'paper'</code> or
do I fix <code>axref='paper'</code> and update the behaviour for
<code>axref='/[xyz][2-9][0-9]/</code> domain' or do I just make
<code>axref='/[xyz][2-9][0-9]/</code> domain' work like <code>axref='/[xyz][2-9][0-9]/'?</code>.</p>

alexcjohnson commented on src/plots/cartesian/constants.js:
```diff
@@ -26,6 +26,7 @@ module.exports = {
     // note that this is more permissive than counterRegex, as
     // id2name, name2id, and cleanId accept "x1" etc
     AX_ID_PATTERN: /^[xyz][0-9]*$/,
+    AX_ID_EXTRAS_PATTERN: /^[xyz]([2-9]|[1-9][0-9]+)?( domain)?$/,
```
The way you use this, it ends up replacing x- or y-specific patterns with a generic xyz pattern. See if you can make something more precise like `idRegex.xExtra` - looks like [`counterRegex`](https://github.com/plotly/plotly.js/blob/6b6b475acbc3ff28a0d5603b16583d2444021730/src/lib/regex.js#L21) has a 2nd argument that will allow this, something like `counterRegex('x', '( domain)?')`

nicholas-esterer replied:
:+1:

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -82,21 +82,92 @@ axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption
+    if(axNumMatch) {
+        var axNum = axNumMatch[1];
+        coerceRefExtras = coerceRefExtras.concat(
+                    (axNum !== undefined) ? [axLetter + axNum + ' domain'] : []);
```
This approach doesn't tell you that `axLetter + axNum` is a valid axis - you could supply `xref = 'x99999 domain'` and it would be accepted.

`axes.coerceRef` is sufficiently specialized, I'd just give it another arg (maybe even just a boolean) to allow domain refs. Then after making `axlist` you can just duplicate each item plus `' domain'`.

nicholas-esterer replied:
:+1:

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -82,21 +82,92 @@ axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption
+ * ar: the axis reference string
+ *
+ */
+axes.extractAxisFromAxisRef = function(ar) {
```
The terminology I think we use elsewhere is:
- "axis id" - a string like `'x'`, `'y2'` used to refer to a specific axis
- "axis name" - a string like `'xaxis'`, `'yaxis2'` that's also the container name in `layout`
- "axis" - the `_fullLayout` object

"axis reference" we perhaps haven't settled on a meaning yet, but I'd suggest using it to mean anything that can be set as the value of an `xref` or `yref` attribute. Then this utility might belong in [cartesian/axis_ids.js](https://github.com/plotly/plotly.js/blob/master/src/plots/cartesian/axis_ids.js) as something like `exports.ref2id` to match the other function names there.

Also you're always feeding it the output of a `coerceRef` call, so you know the input is always going to be a valid axis reference - so you should be able to simplify this to something like "if the first character is an x/y/z, return everything before the first space; otherwise return false" (I feel like `false` would be more useful in the `'paper'` case than still returning `'paper'`, would you agree?)

nicholas-esterer replied:
:+1:

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -82,21 +82,92 @@ axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption
+ * ar: the axis reference string
+ *
+ */
+axes.extractInfoFromAxisRef = function(ar) {
```
You refer to this as "info" or "opt" in the code but those are a bit too generic - perhaps this can be "reference type" (ie make variables like `xRefType` and this function can be called something like `getRefType`) and its return values could be `'paper'`, `'range'`, or `'domain'`?

nicholas-esterer replied:
:+1:

alexcjohnson commented on src/plot_api/helpers.js:
```diff
@@ -217,7 +217,7 @@ exports.cleanLayout = function(layout) {
     var valIn = container[attr];
     var axLetter = attr.charAt(0);
-    if(valIn && valIn !== 'paper') {
+    if(valIn && valIn !== 'paper' && !valIn.endsWith('domain')) {
```
I don't believe we polyfill `endsWith` (for IE11) so we shouldn't use it - here and a few other places too.

Also in principle we should extend `cleanId` to support domain references - so `'x1 domain'` would get converted to `'x domain'`. This would be best as an option to `cleanId` since domain references are only allowed in certain places.

nicholas-esterer replied:
Ok :+1: It looks like the place where `cleanId` could see axis IDs ending with ` domain` is in the places where `cleanAxRef` is used in `src/plot_api/helpers.js`, so I've added an argument to `cleanId` for allowing "domain" type axis references, and set this argument when calling `cleanId` in `cleanAxRef`.
But I noticed that cleanAxRef is not used for images? ... confirming this ...

nicholas-esterer replied:
Yes, indeed if you set a funny axis reference like xref: 'x1' for images, the image doesn't draw. So I've added `cleanAxRef` for images into `src/plot_api/helpers.js` by following what is done there for shapes.

nicholas-esterer replied:
:+1:

alexcjohnson commented on src/constants/axis_placeable_objects.js:
```diff
@@ -0,0 +1,30 @@
+module.exports = {
+    axisRefDescription: function (axisname,lower,upper) {
+        return [
+            `If set to a ${axisname} axis id (e.g. *${axisname}* or`,
```
Nice abstraction! But again we need to downgrade to ES5 - ie just concatenate strings, no backtick templates.

nicholas-esterer replied:
:+

alexcjohnson commented on src/components/annotations/draw.js:
```diff
@@ -73,6 +73,26 @@ function drawOne(gd, index) {
+// the plot.
+// axDomainRef: if true and axa defined, draws relative to axis domain,
+// otherwise draws relative to data (if axa defined) or paper (if not).
+function p2rR2p(axa, optAx, dAx, gsDim, vertical, axDomainRef) {
```
If I'm understanding this correctly, you aren't _converting_ per se, you're shifting a position value (which can be an axis data value, an axis domain fraction, or a paper fraction) to a new position, offset by `dAx` pixels, right? Then I'd suggest renaming the function and args something like:

`function shiftPosition(ax, pos, pxShift, gsDim, vertical, axDomainRef)`

nicholas-esterer replied:
:+1:

alexcjohnson commented on src/components/annotations/defaults.js:
```diff
@@ -65,7 +68,7 @@ function handleAnnotationDefaults(annIn, annOut, fullLayout) {
             // TODO: sometime it might be interesting to allow it to be on *any* axis
             // but that would require updates to drawing & autorange code and maybe more
-            if(aaxRef !== 'pixel' && aaxRef !== axRef) {
+            if(aaxRef !== 'pixel' && aaxRef !== axRefAxOnly) {
```
At some point it may be nice to allow annotation arrow tail references to be axis domains too, but that doesn't need to be now. Until then, I think we need `aaxRef` to still match the complete `axRef` or revert to `'pixel'`.

alexcjohnson replied:
Re: your comment above about these attributes - you're right that specifying the annotation tail in anything other than pixels is fairly limited, ie it doesn't currently support `'paper'` at all. That happens in the `coerceRef` immediately above, where `'paper'` is not allowed as an extra, only `'pixel'`. Again, doesn't need to be added now - we just need to ensure that at the end of `supplyDefaults` we only have valid combinations of attributes given the currently implemented subset of features, regardless of what values the user feeds in.

nicholas-esterer replied:
In the end I made it so that as long as `aaxRef === `axRef` (e.g., `xref === axref`), then it will draw both the annotation and the arrow using that coordinate system. However for all cases other than `pixel`, the coordinates of the arrow head are absolute, which is consistent with how it worked for data references (e.g., `xref === xref === 'x'`). The reason for this is that it made the logic actually weirder because we couldn't just do `if(aaxRef !== 'pixel' && aaxRef !== axRef)`, and I thought it would make users happy :) because it will be consistent with how it works for data coordinates.

nicholas-esterer replied:
:+1:

alexcjohnson commented on src/components/annotations/calc_autorange.js:
```diff
@@ -34,10 +34,12 @@ function annAutorange(gd) {
-        if(xa) calcAxisExpansion(ann, xa);
-        if(ya) calcAxisExpansion(ann, ya);
+        if(xa && (xrefInfo !== 'domain')) calcAxisExpansion(ann, xa);
+        if(ya && (yrefInfo !== 'domain')) calcAxisExpansion(ann, ya);
```
Here's a case where it would be simpler if we had `xRefType = ('paper' | 'range' | 'domain')` - then this could look like:
```js
if(xRefType === 'range') calcAxisExpansion(ann, Axes.getFromId(gd, ann.xref));
```

nicholas-esterer replied:
:+1:

nicholas-esterer commented:
Thanks for your comments, they helped me understand the logic of the code more, and so I tried to make my changes more consistent.
I've addressed all the points you've raised above, however I still need to move the tests in `test/domain_ref_tests.html` over to something working with `jasmine`.
I have a a question about how to make the tests synchronous:
I would like to assert some things about the placement of elements on the page, then change the layout and do this again. My knowledge of asynchronous `Javscript` is limited, but from what I read `jasmine` is asynchronous (I remember trying to delay a page update using `window.setTimeout` within `jasmine` and failing; it seemed to not wait for the timeout).

User interaction:
There were some things that initially looked fine with my PR but when I set `{editable: true}` and started dragging, funny things happened. I've fixed all the funny things I found, and would like to make sure they stay un-funny. So I need to write some interaction tests à la `test/jasmine/tests/parcats_test.js`.

alexcjohnson commented:
Jasmine does a very nice job with async tests in fact - and given that in general our plotting functions can be async, this is used a lot in our tests. A good encapsulated use of this (and close to the area you're working in) is the test below. We use this pattern a lot: we create the `gd` container in `beforeEach` and destroy it in `afterEach`, then to make the test itself async, the test function takes an arg which we call `done`, then the test body looks like:
```js
Plotly.newPlot(gd, fig)
.then(<assertions>)
.then(<modify the plot eg with Plotly.react>)
.then(<more assertions>)
.then(<repeat as needed>)
.catch(failTest)
.then(done)
```

https://github.com/plotly/plotly.js/blob/3781f482455e7623d1a0368b6e74427df549550b/test/jasmine/tests/shapes_test.js#L527-L564

nicholas-esterer commented:
OK, great, thank you for that :)
This is my proposal for a test suite, let me know if it misses some things or goes overboard:
Shapes will be drawn with all combinations of xref|yref=range|domain|paper for both linear and log axes (that's 36 combinations, generated programmatically of course).
The way I thought of testing the shape is to draw a scatter trace, get the bounding box using some SVG utilities that already exist in test/jasmine/assets, and compare it to the drawn shape. It will be easy to draw the exact same shape using a scatter trace if we set the domain and range of the plot axes to known values.
I also want to make sure that the shape goes to an expected location after moving it when editable=true.
Finally I will test the annotations by comparing them with lines drawn using scatter trace, and the images by comparing them with rects also drawn with scatter trace. I wasn't thinking of testing the text.
The shapes tested will be limited to lines and rects, although it's possible to draw more. I thought this would be sufficient to get the plotly.py hrect/vect/hline/vline PR through. Then later we could test ellipses etc., or just assume if the bounding box is correct, then the shape is correct.

alexcjohnson commented:
That all sounds reasonable. The one piece I'd change is: just set the plot size, axis domains, and margins to convenient values, then you should be able to calculate the shape positions precisely without having to reference a scatter trace. For example you can set all the margins to 100, width and height each 700, and axis domains `[0, 0.4]` and `[0.6, 1]`, then each axis has a length exactly 200px, starting at 100 or 400 px from the edge.

alexcjohnson commented:
> But to make this pass I'd have to change an already existing test.

It's normal for low-level tests like this to need changes from time to time. As long as the updated test serves the same purpose with respect to the external API (in this case the output of `PlotSchema.get()`) it doesn't matter where it gets the other side of the equation.

nicholas-esterer commented:
Is there a policy on including a new npm package for testing? I would like to use `itertools` https://www.npmjs.com/package/itertools

alexcjohnson commented:
Include any packages you want for testing purposes - but if they're used in tests only, not in `src`, put them in `package.json` in `devDependencies` rather than `dependencies`.

If you add new packages to `src` (and thus to `dependencies`) we'll scrutinize those a bit more carefully, as they make their way into the final bundles and thus security and file size concerns come into play.

nicholas-esterer commented:
OK! All the tests are passing and I've written tests for these situations:
- Every possible combination of axis reference ([xy][0-9][2-9]+ | "paper" | "[xy][0-9][2-9]+ domain"), every axis type (log or linear) and every possible object (shape, annotation, image). Maybe this test is overkill, but we can easily limit some of the combinations if they seem redundant (just use `filter` on the list of combinations or something).
See `test/jasmine/tests/domain_ref_test.js`
- Some simple editing cases where the user drags shapes or annotations
See `test/jasmine/tests/domain_ref_interact_test.js`
- An image test where the objects are plotted in the correct positions
See `test/image/mocks/domain_refs.json`

archmoj commented on package.json:
```diff
@@ -75,6 +75,7 @@
     "d3-time-format": "^2.2.3",
     "delaunay-triangulate": "^1.1.6",
     "es6-promise": "^4.2.8",
+    "extra-iterable": "^2.5.12",
```
This module shouldn't end up in our dependencies; whereas it is only used in new tests. 
To move it from `dependencies` to `dev-dependencies` you could run the commands below which would install the latest version `2.5.13`:
```
npm uninstall extra-iterable
npm install --save-dev extra-iterable
```

archmoj commented on .gitignore:
```diff
@@ -6,6 +6,8 @@ build/*
 npm-debug.log*
 *.sublime*
+*~
+tags
```
Could you elaborate why we should add these two?

nicholas-esterer replied:
some text editors suffix backup files with `~` and use the file `tags` to store information so you can jump to function definitions. But taggers don't work that well with Javascript it seems, so we could also get rid of the `tags`.

archmoj commented on src/components/annotations/attributes.js:
```diff
@@ -12,6 +12,7 @@ var ARROWPATHS = require('./arrow_paths');
 var fontAttrs = require('../../plots/font_attributes');
 var cartesianConstants = require('../../plots/cartesian/constants');
 var templatedArray = require('../../plot_api/plot_template').templatedArray;
+var axisPlaceableObjs = require('../../constants/axis_placeable_objects.js');
```
You could drop `.js` from the end of such requires.
Also I think it would be more consistent with our code base, if you keep that file in `annotations` instead of `constants`.
Then for requiring from `image` and `shapes`, you could point to the attribute (using Lib.extendFlat) or requiring the function located in `annotation`.

archmoj commented on src/constants/axis_placeable_objects.js:
```diff
@@ -0,0 +1,30 @@
+'use strict';
+
+module.exports = {
+    axisRefDescription: function(axisname, lower, upper) {
```
There is only one function in this file.
So let's simply export it like this:
```js
module.exports = function axisRefDescription(axisname, lower, upper) {
```

archmoj commented on src/constants/axis_placeable_objects.js:
```diff
@@ -0,0 +1,30 @@
+        return [
+            'If set to a ', axisname, ' axis id (e.g. *', axisname, '* or',
+            '*', axisname, '2*), the `', axisname, '` position refers to a',
+            '', axisname, ' coordinate. If set to *paper*, the `', axisname, '`',
```
There is no need to put an empty string at the start of this line (as well as some following lines).
So this could be simplified to
```js
axisname, ' coordinate. If set to *paper*, the `', axisname, '`',
```

archmoj commented on src/plots/cartesian/axes.js:
```diff
@@ -84,28 +84,72 @@ function expandRange(range) {
+ *     of possible values for this axis reference.
  */
-axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption) {
+axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption, domainRef) {
```
It appears this function is called from every place using `domainRef` true. 
@nicholas-esterer could you please confirm if that is intended?

nicholas-esterer replied:
Yes, this is because of the previous review where it was established this would be the better way, I guess because then in the future it would be easy to make cases where `domainRef` would not be allowed.

archmoj replied:
Not clear why we are attaching these to the `cartesian` axes objects?!
IMHO, we should drop the `domainRef` argument here and from the callers.
Then move this `coerceRef` and `addAxRefDomainCoerceRefExtra` functions to `src/components/annotations/axis_ref.js` file.

archmoj replied:
Also this item is not resolved yet.

archmoj replied:
> Not clear why we are attaching these to the `cartesian` axes objects?!
> IMHO, we should drop the `domainRef` argument here and from the callers.
> Then move this `coerceRef` and `addAxRefDomainCoerceRefExtra` functions to `src/components/annotations/axis_ref.js` file.

This is not correct. These functions are in the `axes` module. So let's keep where they are.

archmoj replied:
> Yes, this is because of the previous review where it was established this would be the better way, I guess because then in the future it would be easy to make cases where `domainRef` would not be allowed.

Please only drop `domainRef` from the function argument and the callers since it has been called using `true` from everywhere. Thanks.

archmoj commented on src/plots/cartesian/axes.js:
```diff
@@ -84,28 +84,72 @@ function expandRange(range) {
-    if(!dflt) dflt = axlist[0] || extraOption;
+    if(!dflt) dflt = axlist[0] || (typeof extraOption === 'string' ? extraOption : extraOption[0]);
     if(!extraOption) extraOption = dflt;
+    if(domainRef) axlist = axlist.concat(axlist.map(function(x) { return x + ' domain'; }));
```
`concat`ing an item to the end of an array could be slower than using `push`.
You may rewrite 
```js
axlist = axlist.concat(axlist.map(function(x) { return x + ' domain'; }));
```
as
```js
axlist.push(axlist.map(function(x) { return x + ' domain'; }));
```

archmoj commented on src/plots/cartesian/axes.js:
```diff
@@ -84,28 +84,72 @@ function expandRange(range) {
+ * ar: the axis reference string
+ *
+ */
+axes.getRefType = function(ar) {
```
Non-blocking but one may rewrite this block
```js
return (
  ar === 'paper' || 
  ar === 'pixel' ||
  ar === undefined
) ? ar : /( domain)$/.test(ar) ? 'domain' : 'range';
```
to reduce the if statements and return calls.

nicholas-esterer replied:
We decided in the previous review that the way it is might be the clearest.

archmoj replied:
Hmm... one could simply drop the `else` and simply `return` in the end.
Could you possibly share the link to the conversation?

archmoj commented on src/plot_api/helpers.js:
```diff
@@ -174,6 +174,16 @@ exports.cleanLayout = function(layout) {
         cleanAxRef(shape, 'yref');
     }
 
+    var imagesLen = Array.isArray(layout.images) ? layout.images.length : 0;
```
`Array` is global; so we could drop the `Array.` as follow:
```js
var imagesLen = isArray(layout.images) ? layout.images.length : 0;
```

archmoj commented on src/plots/cartesian/axis_ids.js:
```diff
@@ -18,7 +18,7 @@ var constants = require('./constants');
 exports.id2name = function id2name(id) {
     if(typeof id !== 'string' || !id.match(constants.AX_ID_PATTERN)) return;
-    var axNum = id.substr(1);
+    var axNum = id.split(' ')[0].substr(1);
```
Hmm... not sure about this one.
What happens if one used spaces in the axis name?
We should still be able to handle that case.

nicholas-esterer replied:
you mean something like `'x 2'`? I think that would get caught by the reference coercion, no? IIRC @alexcjohnson  was saying that here we could assume `id` was valid and had already been "cleaned".

archmoj commented on src/plots/cartesian/axis_ids.js:
```diff
@@ -30,12 +30,21 @@ exports.name2id = function name2id(name) {
     if(axLetter && id.charAt(0) !== axLetter) return;
-
-    var axNum = id.substr(1).replace(/^0+/, '');
+    if(/( domain)$/.test(id) && (!domainId)) return;
```
Could we store the result of `/( domain)$/.test(id)` in a variable and reuse it in other places in this function?

archmoj commented on src/plots/cartesian/axis_ids.js:
```diff
@@ -30,12 +30,21 @@ exports.name2id = function name2id(name) {
+    if(/( domain)$/.test(id) && domainId) {
+        return id.charAt(0) + axNum + ' domain';
+    }
     return id.charAt(0) + axNum;
```
Non-blocking: we could combine these two last returns:
```js
var hasDomain = /( domain)$/.test(id);
...
return id.charAt(0) + axNum + (
  (hasDomain && domainId) ? ' domain' : ''
);
```

archmoj commented on src/plots/cartesian/axis_ids.js:
```diff
@@ -82,6 +91,8 @@ exports.listIds = function(gd, axLetter) {
 exports.getFromId = function(gd, id, type) {
     var fullLayout = gd._fullLayout;
+    // remove "domain" suffix
+    id = ((id === undefined) || (typeof(id) !== 'string')) ? id : id.replace(/ *domain/, '');
```
Since we are only interested to substitute the first acurance of ` domain`,
could we simply use `id.replace(' domain', '');` instead of the regular expression form here?

archmoj commented on src/plots/cartesian/axis_ids.js:
```diff
@@ -123,3 +134,20 @@ exports.getAxisGroup = function getAxisGroup(fullLayout, axId) {
+    if(/^[xyz]/.test(ar)) {
+        return ar.split(' ')[0];
+    }
+    return false;
```
Again we may consider combining these two returns and drop the if block.
```js
return /^[xyz]/.test(ar) ? ar.split(' ')[0] : false;
```

archmoj commented on src/components/annotations/draw.js:
```diff
@@ -73,6 +73,26 @@ function drawOne(gd, index) {
+        }
+    } else {
+        return optAx + (vertical ? -dAx : dAx) / gsDim;
+    }
```
To improve performance, we may consider refactoring:
```js
var delta = vertical ? -dAx : dAx;
return !axa ? optAx + delta / gsDim :
    axDomainRef ?
        optAx + delta / axa._length : // here optAx normalized to length of axis (e.g., normally in range 0 to 1). But dAx is in pixels. So we normalize dAx to length of axis before doing the math.
        axa.p2r(axa.r2p(optAx) + dAx);
```

nicholas-esterer replied:
I think inline-ifs are good if they make the code clearer, but here I would not say they do that haha :-)

archmoj commented on src/components/annotations/draw.js:
```diff
@@ -320,10 +341,18 @@ function drawRaw(gd, options, index, subplotId, xa, ya) {
+                        basePx = ax._offset + ax._length * alignPosition;
+                    } else {
+                        basePx = gs.l + gs.w * alignPosition;
+                    }
```
Let's store and reuse 
```js
var hasDomain = axRefType === 'domain';
```
on top.
Then you could avoid if statements for the setter e.g.
```js
basePx = hasDomain ?
  ax._offset + ax._length * alignPosition :
  gs.l + gs.w * alignPosition;
```

archmoj commented on src/components/annotations/draw.js:
```diff
@@ -340,8 +369,29 @@ function drawRaw(gd, options, index, subplotId, xa, ya) {
+                    if(tailRefType === 'domain') {
+                        if(axLetter === 'y') {
+                            arrowLength = 1 - arrowLength;
+                        }
```
We could avoid second repetition by moving `arrowLength = 1 - arrowLength;` block(s) one step up.
```js
var isY = axLetter === 'y';
if(isY) arrowLength = 1 - arrowLength;
if(tailRefType === 'domain') {
  posPx.tail = ax._offset + ax._length * arrowLength;
} else if(tailRefType === 'paper') {
  posPx.tail = isY ? 
    gs.t + gs.h * arrowLength :
    gs.l + gs.w * arrowLength;
} ...

nicholas-esterer replied:
Wouldn't this make `arrowLength` wrong in the case that `tailRefType` is not `domain` or `paper`?

archmoj replied:
Good call. Resolving.

archmoj commented on src/components/images/draw.js:
```diff
@@ -132,8 +133,25 @@ module.exports = function draw(gd) {
+            }
+        } else {
+            width = d.sizex * size.w;
+        }
```
Lets' keep this ternary operators:
```
var width = !xa ? d.sizex * size.w : 
  typeof(d.xref) === 'string' && d.xref.endsWith(' domain') ? 
    xa._length * d.sizex :
    Math.abs(xa.l2p(d.sizex) - xa.l2p(0));

var height = !ya ? d.sizey * size.h :
  typeof(d.yref) === 'string') && d.yref.endsWith(' domain') ?
    ya._length * d.sizey :
    Math.abs(ya.l2p(d.sizey) - ya.l2p(0));
```

archmoj commented on src/components/images/draw.js:
```diff
@@ -142,8 +160,29 @@ module.exports = function draw(gd) {
+            }
+        } else {
+            xPos = d.x * size.w + size.l;
+        }
```
Also here let's try keeping those ternary operators to avoid `if/else` blocks.
```js
var xPos = !xa ? d.x * size.w + size.l :
  typeof(d.xref) === 'string' && d.xref.endsWith(' domain') ?
    xa._length * d.x + xa._offset :
    xa.r2p(d.x) + xa._offset;

xPos += xOffset;
```

archmoj commented on src/components/annotations/draw.js:
```diff
@@ -608,15 +663,22 @@ function drawRaw(gd, options, index, subplotId, xa, ya) {
                 moveFn: function(dx, dy) {
                     var csr = 'pointer';
+                    var xRefType = Axes.getRefType(options.xref);
+                    var yRefType = Axes.getRefType(options.yref);
```
Could we possibly move this test to `shiftPosition` function?

nicholas-esterer replied:
fixed by 2d25857bbe5ee314f7da07669ca4158f072bc831

archmoj commented on src/components/shapes/helpers.js:
```diff
@@ -79,13 +85,21 @@ exports.getDataToPixel = function(gd, axis, isVertical) {
+                return ((isVertical ?
+                        (1 - (p - axis._offset) / axis._length) :
+                        (p - axis._offset) / axis._length));
+            };
```
Let's use a temporary variable to avoid repetition.
```js
pixelToData = function(p) {
    var q = (p - axis._offset) / axis._length);
    return isVertical ? (1 - q) : q;
};
```

nicholas-esterer replied:
d70f4e9c79bb1e8bd5a28002847c60b502da6884 fixes

archmoj commented on test/image/mocks/domain_refs.json:
```diff
@@ -0,0 +1,221 @@
+{
```
Nice image test!
Would you please this new mock to `test/jasmine/tests/mock_test.js` and ensure the test pass locally
```
npm run test-jasmine -- mock
```

nicholas-esterer replied:
35ab9dad6d345f9020b9fd2e813eca9aa998465d makes `mock_test` pass

archmoj commented on test/jasmine/assets/get_rect_center.js:
```diff
@@ -35,13 +37,3 @@ function getRectScreenCoords(rect) {
-
-function findParentSVG(node) {
-    var parentNode = node.parentNode;
-
```
Nice find!

archmoj commented on test/jasmine/assets/svg_tools.js:
```diff
@@ -0,0 +1,25 @@
+    obj.width = svgrect.width;
+    obj.height = svgrect.height;
+    return obj;
+}
```
Great tools!

archmoj commented on test/jasmine/tests/gl2d_plot_interact_test.js:
```diff
@@ -21,7 +21,7 @@ describe('Test removal of gl contexts', function() {
 
     beforeEach(function() {
-        jasmine.DEFAULT_TIMEOUT_INTERVAL = 5000;
+        jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;
```
We may try reverting these (in the end) if we could maintain the performance.

nicholas-esterer replied:
Oh yes! Probably these can be reverted, I doubled them while trying to debug something and forgot to put them 
back.

archmoj replied:
@nicholas-esterer Great! Wanna revert these?

nicholas-esterer replied:
In the end they can't be reverted, the test fail in that case.

archmoj reviewed (commented):
I finished my first round review.
This is a fantastic PR. Thanks @nicholas-esterer :medal_military: for the PR with great tests.
All looking good at this stage. 
And just a few suggestions before we could have a second pass.

archmoj commented on src/plot_api/helpers.js:
```diff
@@ -174,6 +174,16 @@ exports.cleanLayout = function(layout) {
         cleanAxRef(shape, 'yref');
     }
 
+    var imagesLen = isArray(layout.images) ? layout.images.length : 0;
```
```suggestion
    var imagesLen = Array.isArray(layout.images) ? layout.images.length : 0;
```

archmoj commented on src/components/shapes/helpers.js:
```diff
@@ -92,9 +92,10 @@ exports.getPixelToData = function(gd, axis, isVertical, opt) {
     if(axis) {
         if(opt === 'domain') {
             pixelToData = function(p) {
+                var q = p - axis._offset;
```
One could still divide by `axis._length` when computing the temporary variable. Right?
```js
pixelToData = function(p) {
    var q = (p - axis._offset) / axis._length;
    return isVertical ? (1 - q) : q;
};
```

nicholas-esterer replied:
d70f4e9c79bb1e8bd5a28002847c60b502da6884 fixes

nicholas-esterer commented:
> We should also add it on top of this line to be required and tested.
> 
> https://github.com/plotly/plotly.js/blob/f3531dcc28a53450f7f3eeca6a24feef42e2d461/test/jasmine/tests/mock_test.js#L1315

Do you mean below? It is below `display-text_zero-number` in the first list...

archmoj commented:
@alexcjohnson @nicolaskruchten 
I am not convinced that adding ` domain` references to `(x|y)ref` values is a good idea.
Parsing strings in `cartesian/axes` just to make this work for layout objects (not cartesian traces) may not be the best option as it could reduce the performance.
Why not adding a new attribute to control the domain referencing for `shapes`, `annotations` and `images` ?
```
xref: 'x2',
xrefdomain: true
```

alexcjohnson commented:
@archmoj I guess you're concerned about running `/( domain)$/.test(ar)` in `axes.getRefType` multiple times per reference while plotting? In general I feel like these objects are few enough that this isn't going to be a problem. You certainly shouldn't use as many of these as data points.

I do like the look of `xref: 'x2 domain'` - it's compact and intuitive, to my eye anyway. But we could certainly work on performance, and regex tests are definitely the key concern here. By the time we get to `axes.getRefType` we will only have a well-defined set of possibilities for what the reference can be, because we're always looking at the output of `supplyDefaults` rather than direct user input. Maybe the way it's being used it can also be `undefined`, but aside from that `axes.coerceRef` will only give us `paper`, `pixel`, `(x|y)\d+`, or `(x|y)\d+ domain`, right? So something like:
```
axes.getRefType = function(ar) {
    if (!ar || ar === 'paper' || ar === 'pixel') {
        return ar;
    }
    if (ar.charAt(ar.length - 1) === 'n') { // '... domain'
        return 'domain';
    }
    return 'range';
}
```

I see in your comments some concerns about `if` blocks and multiple `return` paths, suggesting converting these to ternary expressions. I'm not aware of a big performance benefit to be had from that kind of optimization but if you have references for this I'm happy to learn more.

The last option I'll mention is that we could run `getRefType` just once during `supplyDefaults`, and store the results in `obj._xrefType` and `obj._xrefId`. That gives us the two separate values to look at, like you're suggesting, without forcing us to give up the combined API we have now. But I'd suggest at least for now leaving this out and just optimizing `axes.getRefType`, I suspect performance that way will be plenty good.

archmoj commented:
@alexcjohnson thanks for the info. 

In fact my main concern is not related to the `getAxType` function which is only called by `shapes`, `annotations` & `images`. Although I'd suggest moving that particular function somewhere in `components`.

On the other hand, changes in `src/plots/cartesian/axis_ids.js` may slow down the interactions.

alexcjohnson commented:
> changes in `src/plots/cartesian/axis_ids.js` may slow down the interactions.

Ah, I see. Thanks for clarifying. Yes, these functions are called a lot, though I notice they already have regex tests in them so they can't be THAT hot. we could certainly simplify them, for example by making `AX_ID_PATTERN` capture the relevant info so no further processing is needed after the `match`
```
AX_ID_PATTERN: /^([xyz])([0-9]*)( domain)?$/

exports.id2name = function id2name(id) {
    if (typeof id !== 'string') return;
    var match = id.match(constants.AX_ID_PATTERN);
    if (!match) return;

    var axNum = match[2];
    if (axNum === '1') axNum = '';
    return match[1] + 'axis' + axNum;
};

archmoj commented on src/constants/axis_placeable_objects.js:
```diff
@@ -0,0 +1,30 @@
+/**
```
This file is not really belong to our `constants`.
Let's move & rename this file to `src/components/annotations/axis_ref.js` and also place `getRefType` into that file please.

nicholas-esterer replied:
Perhaps it shouldn't go in `src/constants` but it is also used by `src/components/images` and `src/components/shapes`, too. So maybe in `src/components` somewhere then?

archmoj replied:
Once placed in `annotations` we would still be able to access it like this from `shapes` and `images`:
```js
var axisRef = require('../annotations/axis_ref');
```

nicholas-esterer replied:
It's true but it's more of a question of respecting the already established organization where everything specific to `annotations` is in its folder etc. I mean why not put it in `src/components/shapes` instead?

archmoj replied:
Good question. `shapes` already reuse parts of the code from `annotations`:
https://github.com/plotly/plotly.js/blob/86f5e2d02d56a94fcf2c26b30d3539fd54211c40/src/components/shapes/attributes.js#L11

archmoj replied:
@nicholas-esterer are you going to address this?

nicholas-esterer replied:
Well yes and no. I can take it out of `'src/constants'`, but I really don't understand what it has to do with `axis_ref` and `getRefType`. This latter function is really related to axes and not at all to documentation strings. The file we are talking about `axis_placeable_objects` generates a documentation string. Also the function has to do with all the `image`, `shape` and `annotation` objects. So I'd like maybe a second opinion on this one.

archmoj commented on src/components/annotations/draw.js:
```diff
@@ -562,19 +610,22 @@ function drawRaw(gd, options, index, subplotId, xa, ya) {
-                            ya.p2r(ya.r2p(options.y) + dy) :
-                            (options.y - (dy / gs.h)));
+                        modifyItem('x',
+                            shiftPosition(xa, options.x, dx, gs.w, false, options.xref));
```
We could possibly refactor `shiftPosition` function in a way here & in other places we simply call it like the following:
```js
shiftPosition(xa, dx, gs, options);
```

archmoj replied:
In order to achieve that the `shiftPosition` arguments may look like this:
```js
function shiftPosition(axa, dAx, gs, opts) {
    var axLetter = axa._id.charAt(0);
    var optAx = opts[axLetter];
    var axRef = opts[axLetter + 'ref'];
    var vertical = axLetter === 'y';
    var gsDim = vertical ? gs.h : gs.w;
```

nicholas-esterer replied:
fixed by 6eba309ac6a535bdf7849c070be56e8f5ac4ec07

archmoj replied:
Resolved thanks to https://github.com/plotly/plotly.js/pull/5014/commits/6eba309ac6a535bdf7849c070be56e8f5ac4ec07 and https://github.com/plotly/plotly.js/pull/5014/commits/759d37bf9e1eac56cef5d1e4df44e899e20ffadc.

archmoj commented on src/components/shapes/helpers.js:
```diff
@@ -58,18 +58,24 @@ exports.extractPathCoords = function(path, paramsToUse) {
 };
 
-exports.getDataToPixel = function(gd, axis, isVertical) {
+exports.getDataToPixel = function(gd, axis, isVertical, opt) {
```
Let's rename this `opt` argument to `refType`.

nicholas-esterer replied:
fixed by 2d25857bbe5ee314f7da07669ca4158f072bc831

archmoj commented on src/components/annotations/draw.js:
```diff
@@ -78,8 +78,12 @@ function drawOne(gd, index) {
+function shiftPosition(axa, options, dAx, axLetter, gs) {
+    var optAx = options[axLetter];
+    var axRef = options[axLetter + 'ref'];
+    var vertical = axLetter.includes('y');
```
Please note that `includes` is not supported in IE.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes

nicholas-esterer replied:
What about `vertical = /y/.test('axLetter')` ?

archmoj replied:
That should work. Or
```js
vertical = axLetter.indexOf('y') !== -1;
```

nicholas-esterer replied:
fixed by 759d37bf9e1eac56cef5d1e4df44e899e20ffadc

archmoj commented on src/components/annotations/draw.js:
```diff
@@ -78,8 +78,12 @@ function drawOne(gd, index) {
 // axDomainRef: if true and axa defined, draws relative to axis domain,
 // otherwise draws relative to data (if axa defined) or paper (if not).
-function shiftPosition(axa, optAx, dAx, gsDim, vertical, axRef) {
+function shiftPosition(axa, options, dAx, axLetter, gs) {
```
Nice simplifications! Thanks.
In plotly.js we often pass `options` as the last argument of the function. And we could read `axLetter` from the first argument. So perhaps the function should look like:
```js
function shiftPosition(axa, dAx, gs, options) {
  var axLetter = axa._id.charAt(0);
```

nicholas-esterer replied:
sometimes `axLetter` is two characters, for example for annotation arrows it is `ay` or `ax`.

archmoj replied:
Good call. So let's keep `axLetter` in the arguments and only move the options to the end. Thanks.

nicholas-esterer replied:
fixed by 759d37bf9e1eac56cef5d1e4df44e899e20ffadc

archmoj commented on src/components/annotations/attributes.js:
```diff
@@ -322,11 +323,7 @@ module.exports = templatedArray('annotation', {
-            'If set to *paper*, the `x` position refers to the distance from',
-            'the left side of the plotting area in normalized coordinates',
-            'where 0 (1) corresponds to the left (right) side.'
+            axisPlaceableObjs.axisRefDescription('x', 'left', 'right'),
```
In addition to `xref` could we use "domain" in `axref` as well?

nicholas-esterer replied:
yes we can, but `xref` has to be exactly the same as `axref` or else `axref` reverts to `pixel`.

archmoj replied:
Cool! Thanks for the info.
So let's update the `axref` and `ayref` as well.

nicholas-esterer replied:
fixed by 759d37bf9e1eac56cef5d1e4df44e899e20ffadc

nicholas-esterer replied:
(sorry wrong review item)

nicholas-esterer commented:
I can't find your comment on this, but I unfortunately couldn't lower `jasmine.DEFAULT_TIMEOUT_INTERVAL` and get all the tests to pass, so I left it the way it was.

archmoj commented:
@nicholas-esterer would you mind fixing the syntax so that test-syntax pass?
To test this locally before each commit you could run 
```
npm run test-syntax
```

archmoj commented on test/image/mocks/domain_ref_axis_types.json:
```diff
@@ -0,0 +1,1767 @@
+      }
+    ],
+    "template": {
+      "data": {
```
Please remove any unused template setup from this mock.

nicholas-esterer replied:
Could we add a git hook like in Plotly.py?

archmoj commented on test/jasmine/assets/domain_ref/components.js:
```diff
@@ -0,0 +1,793 @@
+var getSVGElemScreenBBox = require(
+    '../../assets/get_svg_elem_screen_bbox');
+var SVGTools = require(
+    '../../assets/svg_tools');
```
Non-blocking:
Let's not break these require lines.
What is bad about?
```js
var getSVGElemScreenBBox = require('../../assets/get_svg_elem_screen_bbox');
var SVGTools = require('../../assets/svg_tools');
```

archmoj commented on test/jasmine/assets/domain_ref/components.js:
```diff
@@ -0,0 +1,793 @@
+// xaxistype can be linear|log, only used if xref has type 'range' or 'domain',
+// same for yaxistype and yref
+function annotationTest(gd, layout, x0, y0, ax, ay, xref, yref, axref, ayref,
+    xaxistype, yaxistype, xid, yid) {
```
Non-blocking: 
When having several input arguments to a function like this
```js
function annotationTest(gd, layout, x0, y0, ax, ay, xref, yref, axref, ayref,
    xaxistype, yaxistype, xid, yid) {
```
There is a chance to use wrong order in the caller(s).
It is possible to refactor it to use an object for last arguments. 
Something like:
```js
function annotationTest(gd, layout, opts) {
  var x0 = opts.x0;
  ...
```
Then change the caller to be 
```
annotationTest(gd, layout, {
    x0: x0, 
    y0: y0, 
    ax: ax,
    ay: ay, 
    xref: xref, 
    yref: yref, 
    axref: axref,
    ayref: ayref
    xaxistype: xaxistype, 
    yaxistype: yaxistype, 
    xid: xid, 
    yid: yid
});
```

archmoj commented on test/jasmine/assets/domain_ref/components.js:
```diff
@@ -0,0 +1,793 @@
+
+
+function imageTestCombos() {
+    var testCombos = Array.from(iterable.cartesianProduct(
```
Might be non-blocking:
`Array.from` is not supported in some browsers namely IE11: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from

archmoj commented on test/jasmine/assets/domain_ref/testnumber.js:
```diff
@@ -0,0 +1,3 @@
+// module.exports can be set to a specific test number to run just a single
+// test, otherwise if set to undefined, runs all the tests.
+module.exports = undefined;
```
Sorry. I don't understand if we really need this to be a separate file?

nicholas-esterer replied:
It's there because jasmine can't read environment variables which would be handy to pin point a specific test that is not passing. So instead I made a file that you change. But I can remove this now because the tests pass.

alexcjohnson replied:
@nicholas-esterer have you seen `fit` and `fdescribe`? https://davidtang.io/2016/01/03/controlling-which-tests-run-in-jasmine.html

nicholas-esterer replied:
Thanks, those are handy but in this case I had a for loop that generates many `it(...)` so if I changed it to `fit(...)` it would disable all the tests. Plus I personally don't like changing the source file, environment variables are more ephemeral and the risk of them getting committed is lower. But thanks for the tip!

archmoj commented on test/jasmine/assets/domain_ref/components.js:
```diff
@@ -0,0 +1,793 @@
+    var aroBBox = shapeToBBox(gd.layout, aro);
+    var ret = compareBBoxes(aroBBox, aroPathBBox);
+    console.log('aroBBox: ' + JSON.stringify(aroBBox));
+    console.log('aroPathBBox: ' + JSON.stringify(SVGTools.svgRectToObj(aroPathBBox)));
```
Please remove or comment these `console.log`s here and in other tests.

nicolaskruchten commented:
OK looks like all the checkboxes are in place, which is great! anything still blocking here @archmoj @alexcjohnson  ?

alexcjohnson commented on src/components/images/draw.js:
```diff
@@ -132,8 +133,21 @@ module.exports = function draw(gd) {
-        var height = ya ? Math.abs(ya.l2p(d.sizey) - ya.l2p(0)) : d.sizey * size.h;
+        var width, height;
+        if(xa !== undefined) {
+            width = ((typeof(d.xref) === 'string') && d.xref.endsWith(' domain')) ?
```
I thought I saw a comment about this earlier but perhaps not... `endsWith` is not supported by IE. Can we use `Axes.getRefType` for these instead?

alexcjohnson replied:
In fact you use `Axes.getRefType` down below (line 201 & 2) so you may as well just call it once up above where you get `xa` and `ya`. In fact given how it's used you could probably just make booleans `xIsDomain` and `yIsDomain`

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -84,28 +84,72 @@ function expandRange(range) {
  * extraOption: aside from existing axes with this letter, what non-axis value is allowed?
  *     Only required if it's different from `dflt`
+ * domainRef: true if ' domain' should be appended to the axis items in the list
+ *     of possible values for this axis reference.
```
`domainRef` disappeared but the docstring entry is still here :)

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -84,28 +84,72 @@ function expandRange(range) {
+ * appended to.
+ *
+ */
+axes.addAxRefDomainCoerceRefExtra = function(container, axLetter, coerceRefExtras) {
```
Looks like this function is no longer used anywhere? 🔪

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -129,12 +173,13 @@ axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption
-    var cleanPos, pos;
-
-    if(axRef === 'paper' || axRef === 'pixel') {
+    var cleanPos, pos, axRefType;
+    axRefType = axes.getRefType(axRef);
```
```suggestion
    var cleanPos, pos;
    var axRefType = axes.getRefType(axRef);
```
Minor style point - list uninitialized vars together, but if you're going to initialize a var just declare and initialize it in one statement on its own line.

alexcjohnson commented on src/plots/cartesian/axes.js:
```diff
@@ -129,12 +173,13 @@ axes.coerceRef = function(containerIn, containerOut, gd, attr, dflt, extraOption
         cleanPos = Lib.ensureNumber;
         pos = coerce(attr, dflt);
     } else {
+        // if axRef is 'range' or undefined we will end up here
```
Won't we fail if we get here with `undefined`? We won't find an `ax` so `ax.fraction2r` will error. In which case the condition above could be simplified to `if(axRefType !== 'range')`.

alexcjohnson commented on test/jasmine/tests/domain_ref_test.js:
```diff
@@ -0,0 +1,62 @@
+            return true;
+        }
+        return i === testNumber;
+    }));
```
This is kind of a funny construction, and it looks like the `filter` functions you're using are identical? Can this just be turned into something like:
```js
['annotations', 'images', 'shapes'].forEach(function(componentName) {
    describe('Test ' + componentName), function() {
        var component = domainRefComponents[componentName];
        ...
    });
});
```

alexcjohnson commented:
@nicholas-esterer this is looking great, nice work! Just one blocking comment (`endsWith`) and a few style points, then it should be good to go from my standpoint.

archmoj commented:
:dancer: _domain_ :dancer:
Looking great. The `noCI` tests are also passed on my machine.
@nicholas-esterer thanks very much for the big contribution.
